// react-native-marked/dist/module/lib/Renderer.js
"use strict";

import React from "react";
import { ScrollView, View, Text, TouchableHighlight, Dimensions } from "react-native";
import MarkedList from "@jsamr/react-native-li";
import Disc from "@jsamr/counter-style/presets/disc";
import Decimal from "@jsamr/counter-style/presets/decimal";
import { Slugger } from "marked";
import { Table, Cell, TableWrapper } from "react-native-table-component";
import MDImage from "./../components/MDImage";
import { onLinkPress } from "../utils/handlers";
import { getTableWidthArr } from "../utils/table";
import MDSvg from "./../components/MDSvg";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
class Renderer {
  slugPrefix = "react-native-marked-ele";
  constructor() {
    this.slugger = new Slugger();
    const {
      width
    } = Dimensions.get("window");
    this.windowWidth = width;
  }
  paragraph(children, styles) {
    return this.getViewNode(children, styles);
  }
  blockquote(children, styles) {
    return this.getBlockquoteNode(children, styles);
  }
  heading(text, styles) {
    return this.getTextNode(text, styles);
  }
  code(text, _language, containerStyle, textStyle) {
    return /*#__PURE__*/_jsx(ScrollView, {
      horizontal: true,
      contentContainerStyle: containerStyle,
      children: /*#__PURE__*/_jsx(View, {
        children: this.getTextNode(text, textStyle)
      })
    }, this.getKey());
  }
  hr(styles) {
    return this.getViewNode(null, styles);
  }
  listItem(children, styles) {
    return this.getViewNode(children, styles);
  }
  list(ordered, li, listStyle, textStyle, startIndex) {
    return /*#__PURE__*/_jsx(MarkedList, {
      counterRenderer: ordered ? Decimal : Disc,
      markerTextStyle: textStyle,
      markerBoxStyle: listStyle,
      startIndex: startIndex,
      children: li.map(node => node)
    }, this.getKey());
  }
  escape(text, styles) {
    return this.getTextNode(text, styles);
  }
  link(children, href, styles) {
    return /*#__PURE__*/_jsx(Text, {
      selectable: true,
      accessibilityRole: "link",
      accessibilityHint: "Opens in a new window",
      onPress: onLinkPress(href),
      style: styles,
      children: children
    }, this.getKey());
  }
  image(uri, alt, style) {
    const key = this.getKey();
    if (uri.endsWith(".svg")) {
      return /*#__PURE__*/_jsx(MDSvg, {
        uri: uri
      }, key);
    }
    return /*#__PURE__*/_jsx(MDImage, {
      uri: uri,
      alt: alt,
      style: style
    }, key);
  }
  strong(children, styles) {
    return this.getTextNode(children, styles);
  }
  em(children, styles) {
    return this.getTextNode(children, styles);
  }
  codespan(text, styles) {
    return this.getTextNode(text, styles);
  }
  br() {
    return this.getTextNode("\n", {});
  }
  del(children, styles) {
    return this.getTextNode(children, styles);
  }
  text(text, styles) {
    return this.getTextNode(text, styles);
  }
  html(text, styles) {
    return this.getTextNode(text, styles);
  }
  linkImage(href, imageUrl, alt, style) {
    const imageNode = this.image(imageUrl, alt, style);
    return /*#__PURE__*/_jsx(TouchableHighlight, {
      accessibilityRole: "link",
      accessibilityHint: "Opens in a new window",
      onPress: onLinkPress(href),
      children: imageNode
    }, this.getKey());
  }
  table(header, rows, tableStyle, rowStyle, cellStyle) {
    const widthArr = getTableWidthArr(header.length, this.windowWidth);
    const {
      borderWidth,
      borderColor,
      ...tableStyleRest
    } = tableStyle || {};
    return /*#__PURE__*/_jsx(ScrollView, {
      horizontal: true,
      children: /*#__PURE__*/_jsxs(Table, {
        borderStyle: {
          borderWidth,
          borderColor
        },
        style: tableStyleRest,
        children: [/*#__PURE__*/_jsx(TableWrapper, {
          style: rowStyle,
          children: header.map((headerCol, index) => {
            return /*#__PURE__*/_jsx(Cell, {
              width: widthArr[index],
              data: /*#__PURE__*/_jsx(View, {
                style: cellStyle,
                children: headerCol
              })
            }, `${index}`);
          })
        }), rows.map((rowData, index) => {
          return /*#__PURE__*/_jsx(TableWrapper, {
            style: rowStyle,
            children: rowData.map((cellData, cellIndex) => {
              return /*#__PURE__*/_jsx(Cell, {
                width: widthArr[cellIndex],
                data: /*#__PURE__*/_jsx(View, {
                  style: cellStyle,
                  children: cellData
                })
              }, `${cellIndex}`);
            })
          }, `${index}`);
        })]
      })
    });
  }
  custom(_identifier, _raw, _children, _args) {
    return null;
  }
  getKey() {
    return this.slugger.slug(this.slugPrefix);
  }
  getTextNode(children, styles) {
    return /*#__PURE__*/_jsx(Text, {
      selectable: true,
      style: styles,
      children: children
    }, this.getKey());
  }
  getViewNode(children, styles) {
    return /*#__PURE__*/_jsx(View, {
      style: styles,
      children: children
    }, this.getKey());
  }
  getBlockquoteNode(children, styles) {
    return /*#__PURE__*/_jsx(View, {
      style: styles,
      children: children
    }, this.getKey());
  }
}
export default Renderer;
